core/arch/x86/mangle.c

```diff
@@ -2943,9 +2949,26 @@ mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
             /* we need the whole spill...restore region to all be marked mangle */
             instr_set_our_mangling(instr, true);
             if (spill) {
+                /* We are making several assumptions here. Firstly, we assume that any
+                 * instruction in the mangling code of any control-flow app instruction
+                 * is always before the actual commit point. This shold be safe to assume
+                 * for any control-flow instruction. Such changes in the mangling code of
+                 * the former will be rolled back if needed if xl8 point is in mangling
+                 * code. We therefore do not mark the rip-rel related restores here as
+                 * 'epilogue'. Secondly, we assume that any instructions in mangled code
+                 * that require xsp adjustment to xl8 app state are instructions that can
+                 * be fully rolled back. We do not mark these restores as 'epilogue'
+                 * either.
+                 */
+                instr_t *restore = instr_create_restore_from_tls(
+                    dcontext, scratch_reg, MANGLE_RIPREL_SPILL_SLOT);
+                int xsp_adjust;
                 PRE(ilist, next_instr,
-                    instr_create_restore_from_tls(dcontext, scratch_reg,
-                                                  MANGLE_RIPREL_SPILL_SLOT));
+                    instr_is_cti(instr) ||
```

Comment:
This is how I see this logically: A meta instruction (to be sure that my terminology matches DR language: I am using meta == generated instruction == not app instruction) in the mangling epilogue:
* will be marked with mangling epilogue flag, if it can be translated to the next PC post-app instruction.
* if not marked with mangling epilogue flag, it either is 1) logically not a PC post-app instruction, which is the case for all instr_is_cti. Or 2) it is unsupported to advance to the next PC, and we're making the assumption here that all such instructions can be fully rolled back to the current PC. This is the case e.g. for rip-rel push mangling.

As-is, this would be stable and checked. The way to get into trouble would be if there was a future instruction that needs more complex mangling (e.g. modifies xsp or multiple restores) that can't (or difficult to-) be rolled back (e.g. floating point operation).

Reply:
Comment added.